### PR TITLE
chore: Do not create new memory block when not needed

### LINKED ISSUE #2014: Make dynamic array in acir gen efficient
### Problem

Related to https://github.com/noir-lang/noir/pull/1969#discussion_r1272186521

The way we are generating acir opcodes for dynamic arrays follows too closely to the SSA IR. ie for each array_set, we are creating a new array and then copying over the old array into the new array and then setting just one value in that array at the specified index.

### Happy Case

A rough solution would see an array_set and change the pointer to look at the new array's ID as the pointer for the new array and then only modify one value in that array.

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Array_set was always creating a new copy of an acir MemoryBlock, because it creates a new ssa array

Resolves #2014

## Summary\*

We implement a new ssa pass which maps ssa arrays with their last use.
In acir-gen, when we process the instruction corresponding to the last use of an array, we can avoid to create a new one and copy it, and we can directly operate on that array.
In order to do this, we take benefit from the ssa array-> acir block Id map and map the new ssa array to the block ID of the original ssa array.

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_evaluator/src/ssa/opt/array_use.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+impl Ssa {
+    /// Map arrays with the last instruction that uses it
+    /// For this we simply process all the instructions in execution order
+    /// and update the map whenever there is a match
+    pub(crate) fn array_use(&self) -> HashMap<ValueId, InstructionId> {
+        let mut array_use = HashMap::new();
+        for func in self.functions.values() {
+            let mut reverse_post_order = Vec::new();
+            reverse_post_order.extend_from_slice(PostOrder::with_function(func).as_slice());
+            reverse_post_order.reverse();
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/opt/array_use.rs:23]: We can avoid cloning here if we add an `into_vec` method for `PostOrder`

Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/opt/array_use.rs]: Done

File: crates/noirc_evaluator/src/ssa/opt/array_use.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            Instruction::ArrayGet { array, .. } | Instruction::ArraySet { array, .. } => {
+                let array = dfg.resolve(*array);
+                array_def.insert(array, *instruction_id);
+            }
+            Instruction::Call { arguments, .. } => {
+                for argument in arguments {
+                    if matches!(dfg[*argument], Value::Array { .. }) {
+                        let array = dfg.resolve(*argument);
+                        array_def.insert(array, *instruction_id);
+                    }
```
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/opt/array_use.rs:49]: I would have guess that the type would not change, but indeed it is safer to check the same id we are going to insert.

File: crates/noirc_evaluator/src/ssa.rs [hunk 3]
```diff
@@ -62,7 +62,8 @@ pub(crate) fn optimize_into_acir(
             .dead_instruction_elimination()
             .print(print_ssa_passes, "After Dead Instruction Elimination:");
     }
-    ssa.into_acir(brillig, abi_distinctness)
+    let array_use = ssa.array_use();
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa.rs:65]: Can we rename this to something more descriptive like `find_last_array_uses`?

Comment (guipublic) [reply crates/noirc_evaluator/src/ssa.rs]: Done

**Pull request merged**